src/script_opt/CPP/Attrs.cc

```diff
@@ -7,42 +7,53 @@ namespace zeek::detail
 
 using namespace std;
 
-void CPPCompile::RegisterAttributes(const AttributesPtr& attrs)
+shared_ptr<CPP_InitInfo> CPPCompile::RegisterAttributes(const AttributesPtr& attrs)
 	{
-	if ( ! attrs || attributes.HasKey(attrs) )
-		return;
+	if ( ! attrs )
+		return nullptr;
+
+	auto a = attrs.get();
+
+	if ( processed_attrs.count(a) > 0 )
+		return processed_attrs[a];
 
 	attributes.AddKey(attrs);
-	AddInit(attrs);
 
-	auto a_rep = attributes.GetRep(attrs);
-	if ( a_rep != attrs.get() )
+	// The cast is just so we can make an IntrusivePtr.
+	auto a_rep = const_cast<Attributes*>(attributes.GetRep(attrs));
+	if ( a_rep != a )
 		{
-		NoteInitDependency(attrs.get(), a_rep);
-		return;
+		AttributesPtr a_rep_ptr = {NewRef{}, a_rep};
+		processed_attrs[a] = RegisterAttributes(a_rep_ptr);
+		return processed_attrs[a];
 		}
 
 	for ( const auto& a : attrs->GetAttrs() )
-		{
-		const auto& e = a->GetExpr();
-		if ( e )
-			{
-			if ( IsSimpleInitExpr(e) )
-				{
-				// Make sure any dependencies it has get noted.
-				(void)GenExpr(e, GEN_VAL_PTR);
-				continue;
-				}
+		(void)RegisterAttr(a);
 
-			init_exprs.AddKey(e);
-			AddInit(e);
-			NoteInitDependency(attrs, e);
+	shared_ptr<CPP_InitInfo> gi = make_shared<AttrsInfo>(this, attrs);
+	attrs_info->AddInstance(gi);
+	processed_attrs[a] = gi;
 
-			auto e_rep = init_exprs.GetRep(e);
-			if ( e_rep != e.get() )
-				NoteInitDependency(e.get(), e_rep);
-			}
-		}
+	return gi;
+	}
+
+shared_ptr<CPP_InitInfo> CPPCompile::RegisterAttr(const AttrPtr& attr)
+	{
+	auto a = attr.get();
+
+	if ( processed_attr.count(a) > 0 )
```

Comment:
I have mixed feelings about this.  Performance is generally irrelevant for this code (because it's followed by the much more processing-intensive step of firing up the C++ compiler).  That makes me prefer clarity to performance, and to my taste the current idiom is more clear.  (That said, I get it that that changes the more that one is used to the idiom you frame).  Thoughts?

Reply:
I'm inclined to think the other way. Using iterators and checking against `end()` is the more idiomatic way to do this in C++ these days. The performance benefit does exist, even if it's likely small and obviously dwarfed by the later C++ compiler run.

Reply:
Allrighty.  I've pushed changes to switch to the iterator-based idiom where appropriate.  See what you think.